anndata/base.py

```diff
@@ -1561,6 +1513,123 @@ def to_df(self) -> pd.DataFrame:
             X = self._X
         return pd.DataFrame(X, index=self.obs_names, columns=self.var_names)
 
+    def _get_expression(self, use_raw=False, layer=None):
+        """
+        Convenience method for getting expression values with common arguments and error handling.
+        """
+        is_layer = layer is not None
+        if use_raw and is_layer:
+            raise ValueError(
+                "Cannot use expression from both layer and raw. You provided:"
+                "'use_raw={}' and 'layer={}'".format(use_raw, layer)
+            )
+        if is_layer:
+            return self.layers[layer]
+        elif use_raw:
+            if self.raw is None:
+                raise ValueError("This AnnData doesn't have a value in `.raw`.")
+            return self.raw.X
+        else:
+            return self.X
+
+    def obs_array(
+        self, k: str, *, use_raw: bool = False, layer: Optional[str] = None
+    ) -> np.ndarray:
+        """
+        Convenience function for returning a 1 dimensional ndarray of values
+        from `.X`, `.raw.X`, `.layers[k]`, or `.obs`.
+
+        Made for convenience, not performance. Intentionally permissive about
+        arguments, for easy iterative use.
+
+        Params
+        ------
+        k
+            Key to use. Should be in `.var_names` or `.obs.columns`. If `use_raw`,
+            value should be in `.raw.var_names` instead of `.var_names`.
+        use_raw
+            Whether values should be returned from `.raw.X`
+        layer
+            What layer values should be returned from. If `None`, `.X` is used.
+
+        Returns
+        -------
+        A one dimensional nd array, with values for each obs in the same order
+        as `.obs_names`.
+        """
+        if layer == "X":
+            if "X" in self.layers:
+                pass
+            else:
+                # TODO: Raise deprecation warning
+                layer = None
+
+        if k in self.obs:
+            a = self.obs[k].values
+        elif use_raw:
+            a = self.raw[:, k].X
+        else:
+            idx = self._normalize_indices((slice(None), k))
+            a = self._get_expression(layer=layer)[idx]
+        if issparse(a):
+            a = a.toarray().reshape(-1)
+        return a
+
+
+    def var_array(
```

Comment:
I don't think the naming is optimal and I'm also not sure whether we really want this as a public method. If public, I'd think that something like
```
adata.get_var_1dslice()
```
would make it more clear that this is not another data storage attribute but a convenience function that gives access to existing data.

Reply:
I'd find it useful enough to have as a public function, since it'd be good to advertise that there's a solution to having:

```python
if issparse(a):
    a.toarray().reshape(-1)
```

which is all over the place in my analysis code.

I agree there's a better name. `get_1d_obs` or `obs_array_1d`. I'll think about it a bit.

Reply:
Good! :smile:

How come you have this all over your analysis code and I don't? A sparse matrix slice is cast to a 1d array automatically: https://github.com/theislab/anndata/blob/34f4eb63710628fbc15e7050e5efcac1d7806062/anndata/base.py#L965

Reply:
I get this:

```python
from scipy.sparse import random, issparse
issparse(random(10, 10, format="csr")[0, :])
# True
issparse(random(10, 10, format="csr")[:, 0])
# True
```

But I might be working more directly with the sparse matrices.

Reply:
This does remind me of something. Are there any benchmarks for the overhead of subsetting AnnData? It'd be nice if subsetting was lazy enough to have low overhead, but I'm not sure if this is the case.

Reply:
> But I might be working more directly with the sparse matrices.
Scipy sparse matrices obviously maintain their type, `.X` does the cast for you. That's why I linked to `.toarray()` above. :)

Reply:
> This does remind me of something. Are there any benchmarks for the overhead of subsetting AnnData? It'd be nice if subsetting was lazy enough to have low overhead, but I'm not sure if this is the case.

No, there are no benchmarks and you are absolutely right. It should be lazier. I think the subsets on all the arrays, dataframes etc. are fine because they are themselves lazy (that's at least what I expect). But the `deepcopy(adata.uns)` is awful...